### PR TITLE
UnmarshalKey implemented using pure reflection

### PR DESCRIPTION
**Pull request by dustmop**
### What does this PR do?

This version of UnmarshalKey doesn't rely upon the config having an internal `map[string]interface{}` to store data. Instead, it is implemented using reflection only and demonstrates the a node-based view of the config's data model.

### Motivation

Preparation for the new config model

### Describe how to test/QA your changes

Functionality is covered by unit tests

### CODE REVIEW

File: pkg/config/structure/unmarshal.go
Comment (GustavoCaso) [inline pkg/config/structure/unmarshal.go:93]: Nitpick suggestion

```suggestion
	nodeType := invalidNodeType

	if v.Kind() == reflect.Struct {
		nodeType = structNodeType
	}
	if v.Kind() == reflect.Map {
		nodeType = mapNodeType
	}
	if v.Kind() == reflect.Slice {
		nodeType = listNodeType
	}
	if isScalar(v) {
		nodeType = scalarNodeType
	}

	if nodeType == invalidNodeType {
		return nil, fmt.Errorf("could not create node from: %v of type %T and kind %v", v, v, v.Kind())
	}

	return &nodeimpl{typ: nodeType, val: v}
```

File: pkg/config/structure/unmarshal.go
Comment (GustavoCaso) [inline pkg/config/structure/unmarshal.go:94]: Do we want to panic? Should we return `(node, error)` and handle the case of invalid node type?

File: pkg/config/structure/unmarshal.go
Comment (GustavoCaso) [inline pkg/config/structure/unmarshal.go:36]: Nitpick: Should we use a switch statement here?

File: pkg/config/structure/unmarshal.go
Comment (GustavoCaso) [inline pkg/config/structure/unmarshal.go:256]: If understand correctly, `GetString` would return an empty string if leaf can not get the string value. Do we want to store empty string as map values?

File: pkg/config/structure/unmarshal.go
Comment (GustavoCaso) [inline pkg/config/structure/unmarshal.go:203]: YAML supports floats. Do we care about floats for our use case?

File: pkg/config/structure/unmarshal.go
Comment (GustavoCaso) [inline pkg/config/structure/unmarshal.go:307]: We are accessing `souorce.Size()` three times, should hosting the size value on a variable?

File: pkg/config/structure/unmarshal.go
Comment (GustavoCaso) [reply pkg/config/structure/unmarshal.go:203]: We seem to use float values on our configuration template `pkg/config/config_template.yaml`

**Review by GustavoCaso — COMMENTED** [review]
Awesome work. The code is very easy to follow 🎉 

Left a few questions 😄

File: pkg/config/structure/unmarshal.go
Comment (hush-hush) [inline pkg/config/structure/unmarshal.go:79]: Why have interface for each type here ? It seems we will only ever have 1 implementation for each one, no ?

File: pkg/config/structure/unmarshal.go
Comment (hush-hush) [inline pkg/config/structure/unmarshal.go:189]: Sadly for things like yaml or json we need to support more aliasing as Go accept `true t yes y 1 on` when loading boolean.

File: pkg/config/structure/unmarshal.go
Comment (dustmop) [reply pkg/config/structure/unmarshal.go]: This question made me realize a better approach for representing and constructing the nodes. Now each node is always the implementation that matches its type (inner node, leaf node, array node, map node) and `newNode` returns that implementation immediately, instead of carrying around the reflect kind (the `typ` member variable) to do so later. This removes the need for `AsScalar` and `AsList`, now we just use interface type assertions instead.

This plan is mirroring how the new config model will work. So the main idea is using interfaces instead of implementations in order to control how the model's internal data is being accessed. Getting type dispatch to handle node behavior is a nice benefit.

File: pkg/config/structure/unmarshal.go
Comment (dustmop) [reply pkg/config/structure/unmarshal.go]: Now this method returns a different implementation based upon the value's kind.

File: pkg/config/structure/unmarshal.go
Comment (dustmop) [reply pkg/config/structure/unmarshal.go]: Removed all panics from the file, returning errors everywhere now.

File: pkg/config/structure/unmarshal.go
Comment (dustmop) [reply pkg/config/structure/unmarshal.go]: Yes, good call, done.

File: pkg/config/structure/unmarshal.go
Comment (dustmop) [reply pkg/config/structure/unmarshal.go]: I was using `nil` to mean "no children" before, but now this function returns an error instead in that case.

File: pkg/config/structure/unmarshal.go
Comment (dustmop) [reply pkg/config/structure/unmarshal.go]: Changed this call to `GetString` so that we check the error return. This call isn't quite a conversion: the var `scalar` represents a leaf node, and this call is getting the string value of that leaf node.

File: pkg/config/structure/unmarshal.go
Comment (dustmop) [reply pkg/config/structure/unmarshal.go]: Added float support to the various places that needed it.

File: pkg/config/structure/unmarshal.go
Comment (dustmop) [reply pkg/config/structure/unmarshal.go]: Sure, done, now `numElems` is used for this value.

File: pkg/config/structure/unmarshal.go
Comment (dustmop) [reply pkg/config/structure/unmarshal.go]: Done.

File: pkg/config/structure/unmarshal.go
Comment (dustmop) [reply pkg/config/structure/unmarshal.go]: Yes, I think there's enough in this PR. Expanded on the TODO to specify that support for other maps will be in another PR.

File: pkg/config/structure/unmarshal.go
Comment (dustmop) [reply pkg/config/structure/unmarshal.go]: Oops I think I was spelling "interface" as "inf" which actually reads like "infinity". Regardless, the variable has been eliminated.

File: pkg/config/structure/unmarshal.go
Comment (dustmop) [reply pkg/config/structure/unmarshal.go]: Changed to `mapKeys`. Note that `vals` was being used because these are `reflect.Value` rather than normal strings.

File: pkg/config/structure/unmarshal.go
Comment (dustmop) [reply pkg/config/structure/unmarshal.go]: Do we want / need to support general interface types for keys? It would be nice is the design of the config could move to using strings for keys everywhere, as that simplifies the design a bit and makes it easier to understand. In this specific case, the keys are returned here and then passed to `node.GetChild` which requires a string.

File: pkg/config/structure/unmarshal.go
Comment (dustmop) [reply pkg/config/structure/unmarshal.go]: No longer applicable, as this decision is now handled by the implementation choice.

File: pkg/config/structure/unmarshal.go
Comment (dustmop) [reply pkg/config/structure/unmarshal.go]: Good call, now a helper function is splitting the specifier from the struct tag.

File: pkg/config/structure/unmarshal.go
Comment (dustmop) [reply pkg/config/structure/unmarshal.go]: Done. `fieldNameToKey` now exists.

File: pkg/config/structure/unmarshal.go
Comment (dustmop) [reply pkg/config/structure/unmarshal.go]: Added the additional values to handle bools from yaml.

File: pkg/config/structure/unmarshal.go
Comment (hush-hush) [reply pkg/config/structure/unmarshal.go]: As discussed we can start like this and expand if needed.

File: pkg/config/structure/unmarshal.go
Comment (hush-hush) [inline pkg/config/structure/unmarshal.go:255]: The YAML/JSON go libs also support `1`/`0` for bools.

File: pkg/config/structure/unmarshal.go
Comment (hush-hush) [inline pkg/config/structure/unmarshal.go:304]: This would mean we could Unmarshal a struct into a map, no ? Should we prevent this and check that the source is also a map ?

Similar question for the other `copy*` func

**Review by hush-hush — APPROVED** [review]
I added a couple of quesiton/nit-picks, but it looks good overall !0

File: pkg/config/structure/unmarshal.go
Comment (dustmop) [reply pkg/config/structure/unmarshal.go:304]: I think this is okay? It matches how the standard Unmarshal family of functions work. We'll keep an eye on it in future PRs, as this new function is used in more places, to see if we should decide to disable this behavior.

File: pkg/config/structure/unmarshal.go
Comment (dustmop) [reply pkg/config/structure/unmarshal.go:255]: Ah makes sense, I'll add that before merging.

**Pull request merged**
